mp2-v1/src/receipt_extraction/leaf.rs

```diff
@@ -0,0 +1,471 @@
+            .data
+            .iter()
+            .flat_map(|t| t.to_array())
+            .collect::<Vec<Target>>();
+        let mut out = Vec::new();
+        out.push(self.size);
+        out.extend_from_slice(&self.address.arr);
+        out.push(self.add_rel_offset);
+        out.extend_from_slice(&self.event_signature.arr);
+        out.push(self.sig_rel_offset);
+        out.extend_from_slice(&topics_flat);
+        out.extend_from_slice(&data_flat);
+
+        out
+    }
+
+    pub fn verify_logs_and_extract_values<const NODE_LEN: usize>(
+        &self,
+        b: &mut CBuilder,
+        value: &VectorWire<Target, { PAD_LEN(NODE_LEN) }>,
+        status_offset: Target,
+        relevant_logs_offsets: &VectorWire<Target, MAX_LOGS_PER_TX>,
+    ) -> CurveTarget {
+        let t = b._true();
+        let zero = b.zero();
+        let curve_zero = b.curve_zero();
+        let mut points = Vec::new();
+
+        // Enforce status is true.
+        let status = value.arr.random_access_large_array(b, status_offset);
+        b.connect(status, t.target);
+
+        for log_offset in relevant_logs_offsets.arr.arr {
+            // Extract the address bytes
+            let address_start = b.add(log_offset, self.add_rel_offset);
+
+            let address_bytes = value.arr.extract_array_large::<_, _, 20>(b, address_start);
+
+            let address_check = address_bytes.equals(b, &self.address);
+            // Extract the signature bytes
+            let sig_start = b.add(log_offset, self.sig_rel_offset);
+
+            let sig_bytes = value.arr.extract_array_large::<_, _, 32>(b, sig_start);
+
+            let sig_check = sig_bytes.equals(b, &self.event_signature);
+
+            // We check to see if the relevant log offset is zero (this indicates a dummy value)
+            let dummy = b.is_equal(log_offset, zero);
+
+            let address_to_enforce = b.select(dummy, t.target, address_check.target);
+            let sig_to_enforce = b.select(dummy, t.target, sig_check.target);
+
+            b.connect(t.target, address_to_enforce);
+            b.connect(t.target, sig_to_enforce);
+
+            for &log_column in self.topics.iter().chain(self.data.iter()) {
+                let data_start = b.add(log_offset, log_column.rel_byte_offset);
+                // The data is always 32 bytes long
+                let data_bytes = value.arr.extract_array_large::<_, _, 32>(b, data_start);
+
+                // Pack the data and get the digest
+                let packed_data = data_bytes.arr.pack(b, Endianness::Big);
+                let data_digest = b.map_to_curve_point(
+                    &std::iter::once(log_column.column_id)
+                        .chain(packed_data)
+                        .collect::<Vec<_>>(),
+                );
+
+                // For each column we use the `column_id` field to tell if its a dummy or not, zero indicates a dummy.
+                let dummy_column = b.is_equal(log_column.column_id, zero);
```

Comment:
Not a fan of implicit assumption but if that works then good. However if we're using implicit assumption then let's write it in big in the API and make big checks before we run the circuits on it 🙏

Reply:
(the other option is to not use VectorWire but simply adding another target that tells how many real logs we are extracting no?)